File: chainofresponsibility/Chain.java
package com.batonsystems.settlement.constant;


import java.util.List;
import java.util.Map;
import java.util.function.UnaryOperator;

public class Store {
    private final List<Middleware> middlewares;
    private int index = 0;

    public Store(List<Middleware> middlewares) {
        this.middlewares = middlewares;
    }

    String next(String state) {
        if (index == middlewares.size()) {
            return state;
        }
        return middlewares.get(index++).apply(this::next, state);
    }

    String applyMiddleware(String inititalState) {
        this.index = 0;  // reset index
        return next(inititalState);
    }

    public static void main(String... args) {
        Middleware doubleIt = (nxt, state) -> nxt.apply(state + state);
        Middleware appendHello = (nxt, state) -> nxt.apply(state + " Hello");

        String finalValue = new Store(List.of(new StoreLogger(), doubleIt, doubleIt, appendHello))
                .applyMiddleware("a");
        // {initialValue=a, finalValue=aaaa Hello}

        System.out.println(Map.of("finalValue", finalValue));
        // {finalValue=aaaa Hello}
    }
}

@FunctionalInterface
interface Middleware {
    String apply(UnaryOperator<String> next, String state);
}

class StoreLogger implements Middleware {
    @Override
    public String apply(UnaryOperator<String> next, String state) {
        String finalValue = next.apply(state);
        System.out.println(Map.of(
                "initialValue", state,
                "finalValue", finalValue
        ));
        return finalValue;
    }
}

File: chainofresponsibility/chain.ts
type state = string;

type Next = (state: state) => state;

type Middleware = (next: Next) => (state: state) => state;

type ApplyMiddleware = (middlewares: Middleware[]) => (state: state) => state;

const applyMiddleware: ApplyMiddleware = (middlewares) => (initialState) => {
  let index = 0;
  const next: Next = (state) => {
    const middleware = middlewares[index++];
    return typeof middleware === "function" ? middleware(next)(state) : state;
  };
  return next(initialState);
};

/////////////////////////////////////////


const logger: Middleware = (next) => (state) => {
  const finalValue = next(state);
  console.log({ "initial value": state, "final value": finalValue });
  return finalValue;
};

const doubleMiddleware: Middleware = (next) => (state) => next(state + state);

const appendHello: Middleware = (next) => (state) => next(state + " Hello");

const finalValue = applyMiddleware([
  logger,
  doubleMiddleware,
  doubleMiddleware,
  appendHello,
])("a");
// [LOG]: {
//  "initial value": "a",
//  "final value": "aaaa Hello"
// }

console.log({ finalValue });
// [LOG]: {
//   "finalValue": "aaaa Hello"
// }
